probability/windowed_median.cpp

```diff
@@ -0,0 +1,213 @@
+            eraseFromSorted(_window.front());  // Erase from the multi-value BST
+                                               // the window left side value
+            _window
+                .pop_front();  // Pop the left side value from the window - O(1)
+        }
+    }
+
+    /**
+     * @brief Gets the median of the values in the sliding window
+     * @return Median of sliding window. For even window size return the average
+     * between the two values in the middle
+     */
+    float getMedian() const {
+        if (_sortedValues.size() % 2 != 0) {
+            return *_itMedian;  // O(1)
+        }
+        return 0.5f * *_itMedian + 0.5f * *next(_itMedian);  // O(1)
+    }
+
+    /**
+     * @brief A naive and inefficient method to obtain the median of the sliding
+     * window. Used for testing!
+     * @return Median of sliding window. For even window size return the average
+     * between the two values in the middle
+     */
+    float getMedianNaive() const {
+        auto window = _window;
+        window.sort();  // Sort window - O(NlogN)
+        auto median =
+            *next(window.begin(),
+                  window.size() / 2);  // Find value in the middle - O(N)
+        if (window.size() % 2 != 0) {
+            return median;
+        }
+        return 0.5f * median +
+               0.5f * *next(window.begin(), window.size() / 2 - 1);  // O(N)
+    }
+};
+}  // namespace probability
+
+/**
+ * @brief A testing function
+ * @param vals Stream of values
+ * @param windowSize Size of sliding window
+ */
+static void test(const std::vector<int> &vals, int windowSize) {
+    probability::WindowedMedian windowedMedian(windowSize);
+    for (const auto val : vals) {
+        windowedMedian.insert(val);
+
+        // Comparing medians: efficient function vs. Naive one
+        assert(windowedMedian.getMedian() == windowedMedian.getMedianNaive());
+    }
+}
+
+/**
+ * @brief Main function
+ * @param argc commandline argument count (ignored)
+ * @param argv commandline array of arguments (ignored)
+ * @returns 0 on exit
+ */
+int main(int argc, const char *argv[]) {
+    test({1, 2, 3, 4, 5, 6, 7, 8, 9}, 3);
+    test({9, 8, 7, 6, 5, 4, 3, 2, 1}, 3);
+    test({9, 8, 7, 6, 5, 4, 5, 6}, 4);
+    test({3, 3, 3, 3, 3, 3, 3, 3, 3}, 3);
+    test({3, 3, 3, 3, -7, 3, 3, 3, 3}, 3);
+    test({4, 3, 3, -5, 7, 1, 3, 4, 5}, 5);
+    test({470211272, 101027544, 1457850878, 1458777923, 2007237709, 823564440,
+          1115438165, 1784484492, 74243042, 114807987},
+         6);
+    std::srand(static_cast<unsigned int>(std::time(nullptr)));
+    std::vector<int> vals;
+    for (int i = 8; i < 100; i++) {
+        const auto n = 1 + std::rand() / ((RAND_MAX + 5u) / 20);
+        auto windowSize = 1 + std::rand() / ((RAND_MAX + 3u) / 10);
+        vals.clear();
+        vals.reserve(n);
+        for (int i = 0; i < n; i++) {
+            vals.push_back(rand() - RAND_MAX);
+        }
+        test(vals, windowSize);
```

Comment:
Please add some documentation of what this code does here.

Reply:
OK